### PR TITLE
Unified use of query parameters on search endpoints

### LINKED ISSUE #8: Unify support for from, size, filter and text query parameters across all listings
Every listing on the system should support a combination of the following HTTP query parameters:

- **from**: Starting offset
- **size**: Amount of results retrieved
- **filter**: Logical & Comparison expression filter (e.g.: (a = b) && ( c != b) )
- **text**: Full text search query on an object

### PR DESCRIPTION
**Pull request by umbreak**
Unified support for the following query parameters:
from, size, filter and term.

Fixes #8.

### CODE REVIEW

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+  /**
+    * Lists all outgoing ids linked to the if identified by ''id'' that match the given filter.
+    *
+    * @param id         the selected id (this)
+    * @param filter     the filter to apply to outgoing ids
+    * @param pagination the pagination values
+    * @param term       the optional full text search term
+    */
+  def outgoing(id: String, filter: Filter, pagination: Pagination, term: Option[String] = None)(implicit Q: ConfiguredQualifier[Id]): F[QueryResults[Id]] = {
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala]: using strings is quite unsafe in this scenario; make it generic with a qualifier

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    val thisFilter = Filter(ComparisonExpr(Op.Eq, UriTerm("uuid".qualify), LiteralTerm(s""""$id"""")))
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala]: The uuid is specific to instances, I would recommend filtering based on the full id instead.

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+  /**
+    * Lists all incoming ids linked to the id identified by ''id'' that match the given filter.
+    *
+    * @param id         the selected id (this)
+    * @param filter     the filter to apply to incoming id
+    * @param pagination the pagination values
+    * @param term       the optional full text search term
+    */
+  def incoming(id: String, filter: Filter, pagination: Pagination, term: Option[String] = None)(implicit Q: ConfiguredQualifier[Id]): F[QueryResults[Id]] = {
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala]: Same as for outgoing.

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+/**
+  * Collection of queries.
+  *
+  * @param queryClient   the client to use for executing the queries
+  * @param querySettings the default query settings
+  * @tparam F  the monadic effect type
+  * @tparam Id the generic type which defines the response's payload
+  */
+class FilterQueries[F[_], Id](queryClient: SparqlQuery[F], querySettings: QuerySettings) {
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala]: Once we unify indexing under a single namespace we can move the `Id` tparam to each method instead to be able to share the same FilterQueries instance.

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+  def filterFrom(deprecatedOps: Option[Boolean], filterOps: Option[Filter], baseVoc: Uri): Filter =
+    deprecatedAndRev(deprecatedOps, baseVoc) and filterOps.map(_.expr)
+
+  /**
+    * Constructs a new filter based on an optional deprecation parameter.
+    *
+    * @param deprecatedOps the optional deprecated field
+    * @param baseVoc       the nexus core vocabulary base
+    */
+  def deprecatedAndRev(deprecatedOps: Option[Boolean], baseVoc: Uri): Filter =
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilterQueries.scala]: rename `deprecatedOps` to `deprecatedOpt`

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala [hunk 6]
```diff
@@ -47,13 +47,26 @@ trait QueryDirectives {
     }
 
   /**
-    * Extracts the pagination and filter query params from the request or defaults to preconfigured values.
+    * Extracts the ''term'' query param from the request.
+    */
+  def term: Directive1[Option[String]] =
+    parameter('term.?).flatMap(provide(_))
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala]: @MFSY do you have any thoughts on whether changing this query parameter is a good idea or not?
This used to be `q`.

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala [hunk 7]
```diff
@@ -47,13 +47,26 @@ trait QueryDirectives {
...
+
+  /**
+    * Extracts the ''deprecated'' query param from the request.
+    */
+  def deprecated: Directive1[Option[Boolean]] =
+    parameter('deprecated.as[Boolean].?).flatMap(provide(_))
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala]: I would recommend avoiding this expressing functions like this: `provide(_)`.
Use `flatMap(opt => provide(opt))`.

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/DomainRoutes.scala [hunk 8]
```diff
@@ -24,27 +27,25 @@ import scala.concurrent.{ExecutionContext, Future}
...
-  * @param queryBuilder  query builder for domains
-  * @param base          the service public uri + prefix
+  * @param domains           the domain operation bundle
+  * @param domainQueries     query builder for domains
+  * @param base              the service public uri + prefix
+  * @param querySettings     query parameters from settings
+  * @param filteringSettings filtering parameters from settings
   */
-final class DomainRoutes(domains: Domains[Future], querySettings: QuerySettings, queryBuilder: FilterQueries[DomainId], base: Uri) {
+final class DomainRoutes(domains: Domains[Future], domainQueries: FilterQueries[Future, DomainId], base: Uri)(implicit querySettings: QuerySettings, filteringSettings: FilteringSettings) {
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/DomainRoutes.scala]: align the arguments verticaly please.

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/DomainRoutes.scala [hunk 9]
```diff
@@ -24,27 +27,25 @@ import scala.concurrent.{ExecutionContext, Future}
...
     handleRejections(RejectionHandling.rejectionHandler) {
       pathPrefix("organizations" / Segment / "domains") { orgIdString =>
         val orgId = OrgId(orgIdString)
-        pathEndOrSingleSlash {
-          (get & parameter('from.as[Int] ? pagination.from) & parameter('size.as[Int] ? pagination.size) & parameter('deprecated.as[Boolean].?)) { (from, size, deprecated) =>
-            traceName("listDomains") {
-              queryBuilder.listingQuery(orgId, deprecated, None, Pagination(from, size)).response
-            }
+        (pathEndOrSingleSlash & get & searchQueryParams) { (pagination, filterOpt, termOpt, deprecatedOpt) =>
+          traceName("listDomains") {
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/DomainRoutes.scala]: should we rename to `filterDomains` instead? dunno yet. @MFSY ?

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/InstanceRoutes.scala [hunk 10]
```diff
@@ -57,191 +54,151 @@ class InstanceRoutes(
...
-              parameter('deprecated.as[Boolean].?) { deprecated =>
-                val filter = Filter(deprecatedAndRev(deprecated)) and filterOpt.map(_.expr)
-                buildResponse(instanceQueries.list(orgId, filter, pagination), pagination)
-              }
-            }
+      pathPrefix("data") {
+        (pathEndOrSingleSlash & get & searchQueryParams) { (pagination, filterOpt, termOpt, deprecatedOpt) =>
+          traceName("listInstances") {
+            val filter = filterFrom(deprecatedOpt, filterOpt, querySettings.nexusVocBase)
+            instanceQueries.list(filter, pagination, termOpt) buildResponse(base, pagination)
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/InstanceRoutes.scala]: I would recommend avoiding the use of infixes for these function shapes.

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/InstanceRoutes.scala [hunk 11]
```diff
@@ -57,191 +54,151 @@ class InstanceRoutes(
...
-                        }
+            pathPrefix(Segment) { name =>
+              (pathEndOrSingleSlash & get & searchQueryParams) { (pagination, filterOpt, termOpt, deprecatedOpt) =>
+                traceName("listInstancesOfSchemaName") {
+                  val filter = filterFrom(deprecatedOpt, filterOpt, querySettings.nexusVocBase)
+                  instanceQueries.list(domainId, name, filter, pagination, termOpt) buildResponse(base, pagination)
+                }
+              } ~
+              pathPrefix(Segment) { versionString =>
+                Version(versionString) match {
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/InstanceRoutes.scala]: Not necessarily to be addressed in this change, but we should have a version directive with its rejection and handling to simplify routes that involve a version.

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala (hunk 6 above)
Comment (MFSY) [reply modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala]: I would recommend to keep using q as query parameter just like google.com (www.google.com?q=nexus) .

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/DomainRoutes.scala (hunk 9 above)
Comment (MFSY) [reply modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/DomainRoutes.scala]: I like the prefix 'search'. It's independent from the modality (filter, full text search) used to access the data

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala (hunk 6 above)
Comment (umbreak) [reply modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala]: We keep it as `q` then.

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/DomainRoutes.scala (hunk 9 above)
Comment (umbreak) [reply modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/DomainRoutes.scala]: searchDomains
searchOrganizations
searchSchemas
searchInstances

**Comment by bogdanromanx** [comment]
Please use: `Fixes #issuenumber` in the pull request description.

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/InstanceRoutes.scala (hunk 11 above)
Comment (umbreak) [reply modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/InstanceRoutes.scala]: Done

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/InstanceRoutes.scala (hunk 10 above)
Comment (umbreak) [reply modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/InstanceRoutes.scala]: Done

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilteredQuery.scala [hunk 12]
```diff
@@ -64,16 +65,15 @@ object FilteredQuery {
...
-    * @param thisFilter   the filter to be applied to select the subjects
+    * @param thisSubject  the qualified uri of the subject to be selected
     * @param targetFilter the filter to be applied to filter the objects
     * @param pagination   the pagination settings for the generated query
     */
-  def outgoing(thisFilter: Filter, targetFilter: Filter, pagination: Pagination, term: Option[String] = None): String = {
+  def outgoing(thisSubject: Uri, targetFilter: Filter, pagination: Pagination, term: Option[String] = None): String = {
     val where =
       s"""
-         |?ss ?p ?s .
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/query/builder/FilteredQuery.scala]: This line should not be removed from the query.

**Pull request merged**